core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java

```diff
@@ -269,22 +273,23 @@ public NegotiationResult declined(ClaimToken token, String negotiationId) {
      * @return the number of processed ContractNegotiations.
      */
     private int sendContractOffers() {
-        var processes = negotiationStore.nextForState(ContractNegotiationStates.REQUESTING.code(), batchSize);
+        var processes = negotiationStore.nextForState(REQUESTING.code(), batchSize);
 
         for (ContractNegotiation process : processes) {
             var offer = process.getLastContractOffer();
-            var response = sendOffer(offer, process, ContractOfferRequest.Type.INITIAL);
-            if (response.isCompletedExceptionally()) {
-                process.transitionRequesting();
-                monitor.debug(format("[Consumer] Failed to send contract offer with id %s. ContractNegotiation %s stays in state %s.",
-                        offer.getId(), process.getId(), ContractNegotiationStates.from(process.getState())));
-                continue;
-            }
-
-            process.transitionRequested();
-            monitor.debug(String.format("[Consumer] ContractNegotiation %s is now in state %s.",
-                    process.getId(), ContractNegotiationStates.from(process.getState())));
-            negotiationStore.save(process);
+            sendOffer(offer, process, ContractOfferRequest.Type.INITIAL)
+                    .whenComplete((response, throwable) -> {
+                        if (throwable == null) {
+                            process.transitionRequested();
+                            monitor.debug(String.format("[Consumer] ContractNegotiation %s is now in state %s.",
+                                    process.getId(), ContractNegotiationStates.from(process.getState())));
+                            negotiationStore.save(process);
+                        } else {
+                            String message = format("[Consumer] Failed to send contract offer with id %s. ContractNegotiation %s stays in state %s.",
```

Comment:
The line `process.transitionRequesting();` is missing in the else block. It there a reason why?

Reply:
I removed it here as in the other cases because I don't see the point to set the state as it is already (in this case the `save` method was already missing, so setting it to `transitionRequesting` is definitely not useful)

Reply:
We added this for retry cases - that are not yet implemented. The transition method increases the state count when a process enteres the same state repeatedly. This way, we can abort the process as soon as a certain number of retries is reached.

Reply:
got it, I was asking myself what's the scope of stateCount, since it's not used by anyone. should that be addressed in a new issue? In the meantime I will reintroduce "state tick"

Reply:
The state count should increase since it is transitioning back into the same state. As @juliapampus mentioned, we just haven't yet implemented logic for a transition threshold before moving the process into an error state. That should be a configurable threshold. For the TPM, we should discuss if the threshold is the same for all states or not. We also will need to handle this for contract negotiation as well.